extensions/wikia/FilePage/WikiaFilePage.php

```diff
@@ -40,29 +40,96 @@ protected function isVideo() {
 	/**
 	 * Render the image or video
 	 */
-	protected function openShowImage(){
+	protected function openShowImage() {
 		if ( $this->isVideo() ) {
 			$this->openShowVideo();
 		} else {
 			parent::openShowImage();
 		}
 	}
 
+	/**
+	 * Render the image or video
+	 */
+	public function view() {
+		global $wgMemc, $wgDBprefix, $wgOut, $wgUser, $wgGroupPermissionsLocal, $wgNamespaceProtection;
+		if ( !$wgUser->isAnon() ) {
+			parent::view();
+
+			return;
+		}
+		//fallback to main page
+		$url = wfExpandUrl( "/wiki/Main_Page" );
+		//wiki needs read privileges
+		if ( preg_match( "/\*\|read\|0/", $wgGroupPermissionsLocal ) ||
+			 preg_match( "/\user\|read\|0/", $wgGroupPermissionsLocal ) ) {
+			$wgOut->redirect( $url );
+
+			return;
+		}
+		$redirKey = "redir:" . $wgDBprefix . ":" . $this->getTitle();
+		$url = $wgMemc->get( $redirKey );
+		if ( $url ) {
+			$wgOut->redirect( $url );
+		}
+		$displayImg = $img = false;
+		Hooks::run( 'ImagePageFindFile', [ $this, &$img, &$displayImg ] );
+		if ( !$img ) { // not set by hook?
+			$img = wfFindFile( $this->getTitle() );
+			if ( !$img ) {
+				$img = wfLocalFile( $this->getTitle() );
+			}
+		}
+		if ( !$img ) {
+			parent::view();
+
+			return;
+		}
+		$res = $this->fetchLinks( $img->getTitle()->getDBkey() );
+
+		foreach ( $res as $row ) {
+			$title = Title::newFromRow( $row );
+			if ( $title->isRedirect() ) {
+				continue;
+			}
+			$url = wfExpandUrl( "/wiki/" . $row->page_title );
+			$res->free();
+			break;
+		}
+		$wgMemc->add( $redirKey, $url );
+		$wgOut->redirect( $url );
+	}
+
+	/**
+	 * Fetch informationabout pages linked to image
+	 * @param string $dbKey
+	 * @return string
+	 */
+	private function fetchLinks( $dbKey ) {
+		$dbr = wfGetDB( DB_SLAVE );
+
+		return $dbr->select( [ 'imagelinks', 'page' ], [
+			'page_title',
+			'page_namespace',
+		], [ 'il_to' => $dbKey, 'page_is_redirect' => 0, 'il_from = page_id' ], __METHOD__,
```

Comment:
I think we should consider limiting this to article pages only, so `'page_namespace' => NS_MAIN,` rather than redirecting to pages that potentially won't be indexed or aren't valuable. We can double check if this should be the case.

I'd also suggest reformatting this to be a little more in line with similar code elsewhere which is a bit more readable:
```php
return $dbr->select(
	[ 'imagelinks', 'page' ],
	[ 'page_title', 'page_namespace' ],
	[
		'il_to' => $dbKey,
		'page_is_redirect' => 0,
		'il_from = page_id'
	],
	__METHOD__,
	[
		'LIMIT' => 5,
		'ORDER BY' => 'page_namespace, page_id',
	]
);
```